File: hola-mundo-backend-2022-grupo-7/src/main/kotlin/trabajoPracticoAlgoritmos2/Viaje.kt
package trabajoPracticoAlgoritmos2

class Viaje(val itinerario: Itinerario, val vehiculo: Vehiculo,val usuario: Usuario,val diasDeAlquiler:Int = 0){

    val observers = mutableListOf<Observers>()

    fun calcularCosto():Double{
        return  usuario.calcularCostoDeViaje(itinerario.destino) + vehiculo.costoDeAlquiler(diasDeAlquiler) + itinerario.costoDelItinerario()
    }

    fun realizarViaje(){

        if(usuario.puedeRealizarItinerario(itinerario)){
            usuario.agregarDestinoVisitado(this.itinerario.destino)

            observers.forEach { it.evento(usuario,this) }
        }else{
            throw RuntimeException("El usuario no puede realizar el viaje")
        }
    }

    fun agregarObserver(observer: Observers){
        observers.add(observer)
    }

    fun quitarObserver(observer: Observers){
        observers.add(observer)
    }

}

File: Angular/src/Dominio/Viaje/Viaje.ts
import { Itinerario } from "../Itinerarios/Itinerarios";
import { Usuario } from "../Usuario/Usuario";
import { Vehiculo } from "../vehiculo/vehiculo";

class Viaje{

    readonly itinerarios:Array<Itinerario>
    readonly vehiculo:Vehiculo
    readonly usuario:Usuario
    readonly diasDelAlquiler:number

    constructor(_itinerarios:Array<Itinerario>, _vehiculo:Vehiculo, _usuario:Usuario, _diasDelAlquiler:number){

        this.itinerarios = _itinerarios
        this.vehiculo = _vehiculo
        this.usuario = _usuario
        this.diasDelAlquiler = _diasDelAlquiler

    }

}